source/pkcs11.c

```diff
@@ -12,3 +20,359 @@
+        "id=%p: Unloading PKCS#11. C_Finalize:%s",
+        (void *)pkcs11_lib,
+        pkcs11_lib->should_finalize ? "yes" : "omit");
+
+    if (pkcs11_lib->should_finalize) {
+        CK_RV rv = pkcs11_lib->function_list->C_Finalize(NULL);
+        if (rv != CKR_OK) {
+            /* Log about it, but continue cleaning up */
+            s_raise_ck_error(pkcs11_lib, "C_Finalize", rv);
+        }
+    }
+
+    aws_shared_library_clean_up(&pkcs11_lib->shared_lib);
+    aws_mem_release(pkcs11_lib->allocator, pkcs11_lib);
+}
+
+struct aws_pkcs11_lib *aws_pkcs11_lib_new(
+    struct aws_allocator *allocator,
+    const struct aws_pkcs11_lib_options *options) {
+
+    /* Create the struct */
+    struct aws_pkcs11_lib *pkcs11_lib = aws_mem_calloc(allocator, 1, sizeof(struct aws_pkcs11_lib));
+    aws_ref_count_init(&pkcs11_lib->ref_count, pkcs11_lib, s_pkcs11_lib_destroy);
+    pkcs11_lib->allocator = allocator;
+
+    /* Load the library. */
+
+    /* need a null-terminated string to call next function,
+     * or NULL if going to search the current application for PKCS#11 symbols. */
+    struct aws_string *filename_storage = NULL;
+    const char *filename = NULL;
+    if (options->filename.len > 0) {
+        filename_storage = aws_string_new_from_cursor(allocator, &options->filename);
+        filename = aws_string_c_str(filename_storage);
+    }
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_IO_PKCS11,
+        "Loading PKCS#11. file:'%s' C_Initialize:%s",
+        filename ? filename : "<MAIN PROGRAM>",
+        options->omit_initialize ? "omit" : "yes");
+
+    if (aws_shared_library_init(&pkcs11_lib->shared_lib, filename)) {
+        goto except;
+    }
+
+    /* Find C_GetFunctionList() and call it to get the list of pointers to all the other functions */
+    CK_C_GetFunctionList get_function_list = NULL;
+    if (aws_shared_library_find_function(
+            &pkcs11_lib->shared_lib, "C_GetFunctionList", (aws_generic_function *)&get_function_list)) {
+        goto except;
+    }
+
+    CK_RV rv = get_function_list(&pkcs11_lib->function_list);
+    if (rv != CKR_OK) {
+        s_raise_ck_error(pkcs11_lib, "C_GetFunctionList", rv);
+        goto except;
+    }
+
+    /* Call C_Initialize() (skip if omit_initialize is set) */
+    if (!options->omit_initialize) {
+        CK_C_INITIALIZE_ARGS init_args = {
+            /* encourage lib to use our locks */
+            .CreateMutex = s_pkcs11_create_mutex,
+            .DestroyMutex = s_pkcs11_destroy_mutex,
+            .LockMutex = s_pkcs11_lock_mutex,
+            .UnlockMutex = s_pkcs11_unlock_mutex,
+            /* but if it needs to use OS locks instead, sure whatever you do you */
+            .flags = CKF_OS_LOCKING_OK,
+        };
+
+        CK_RV rv = pkcs11_lib->function_list->C_Initialize(&init_args);
+        if (rv != CKR_OK) {
+            s_raise_ck_error(pkcs11_lib, "C_Initialize", rv);
+            goto except;
+        }
+
+        pkcs11_lib->should_finalize = true;
```

Comment:
do we need to call finalize even if initialize fails, to do partial cleanup?

Reply:
huh. good question.

usually in C APIs, if an "init" call fails it's like nothing ever happened, so my instinct would be that it's not necessary

Looking at [SoftHSM's implementation](https://github.com/opendnssec/SoftHSMv2/blob/3de59817e17bd1873f1e802757b5b38a1c7ca898/src/lib/SoftHSM.cpp#L613), if the `C_Initialize()` call fails then `C_Finalize()` will also fail with `CKR_CRYPTOKI_NOT_INITIALIZED`